src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -85,26 +84,63 @@ public Task RunAsync(ActionRunStage stage)
-            cleanOutputsStep.ContextName = ExecutionContext.ContextName;
-            // Use the same reference type as our composite steps.
-            cleanOutputsStep.Reference = Action;
+            try
+            {
+                // This is where we run each step.
+                await RunStepsAsync(compositeSteps);
 
-            var actionRunner2 = HostContext.CreateService<IActionRunner>();
-            actionRunner2.Action = cleanOutputsStep;
-            actionRunner2.Stage = ActionRunStage.Main;
-            actionRunner2.Condition = "always()";
-            ExecutionContext.RegisterNestedStep(actionRunner2, inputsData, location, Environment, true);
+                // This is where we set the outputs.
+                // All steps have the same scope for their ExecutionContext/
+                // So we can use one of them so that we can get the right "outputs" attribute.
+                HandleOutput(compositeSteps?[0].ExecutionContext);
+            }
+            catch (Exception ex)
+            {
+                // Composite StepRunner should never throw exception out.
+                Trace.Error($"Caught exception from composite steps {nameof(CompositeActionHandler)}: {ex}");
+                ExecutionContext.Error(ex);
+                ExecutionContext.Result = TaskResult.Failed;
+            }
+        }
+
+        private void CompleteStep(IStep step, TaskResult? result = null, string resultCode = null)
+        {
+            var executionContext = step.ExecutionContext;
 
-            return Task.CompletedTask;
+            executionContext.Complete(result, resultCode: resultCode);
+        }
+
+        private void HandleOutput(IExecutionContext stepExecutionContext)
+        {
+            ArgUtil.NotNull(stepExecutionContext, nameof(stepExecutionContext));
+
+            // Evaluate the mapped outputs value
+            if (Data.Outputs != null)
+            {
+                // Evaluate the outputs in the steps context to easily retrieve the values
+                var actionManifestManager = HostContext.GetService<IActionManifestManager>();
+
+                // Format ExpressionValues to Dictionary<string, PipelineContextData>
+                var evaluateContext = new Dictionary<string, PipelineContextData>(StringComparer.OrdinalIgnoreCase);
+                foreach (var pair in stepExecutionContext.ExpressionValues)
+                {
+                    evaluateContext[pair.Key] = pair.Value;
+                }
+
+                // Get the evluated composite outputs' values mapped to the outputs named
+                DictionaryContextData actionOutputs = actionManifestManager.EvaluateCompositeOutputs(stepExecutionContext, Data.Outputs, evaluateContext);
+
+                // Set the outputs for the outputs object in the whole composite action
+                actionManifestManager.SetAllCompositeOutputs(ExecutionContext, actionOutputs);
```

Comment:
feels like the responsibility for setting the outputs belongs in this handler (above evaluate function on manifest manager makes sense)